### PR TITLE
Sema: report conformances and typealias underlying types behind non-public imports

### PR DESCRIPTION
**Pull request by xymus**
Extend exportability checking to support non-public imports in some cases. Error on conformances imported from a non-public import is used in API, and on typealiases that desugars to a type behind a non-public import is used in API. Do no use the exportability checking logic for other checks that are already handled by the checks on the access levels.

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: test/Sema/access-level-import-conformances.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+//--- ConformanceDefinition.swift
+import ConformanceBaseTypes
+extension ConformingType : Proto  {}
+
+//--- Client.swift
+public import ConformanceBaseTypes
+internal import ConformanceDefinition
+
+public func useInAPI(a: any Proto = ConformingType()) { // expected-error {{cannot use conformance of 'ConformingType' to 'Proto' here; 'ConformanceDefinition' was not imported publicly}}
```
Comment (tshortli) [inline test/Sema/access-level-import-conformances.swift:26]: Should there be a corresponding `note:` on the import decl?

**Review by tshortli — APPROVED** [review]

**Comment by xymus** [comment]
@swift-ci Please smoke test macOS

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: test/Sema/access-level-import-conformances.swift (hunk 1 above)
Comment (xymus) [reply test/Sema/access-level-import-conformances.swift:26]: Very good point. I've added the note.

**Pull request merged**